### PR TITLE
Separated out much of the functionality in the AbstractDwcControl int…

### PR DESCRIPTION
**Pull request by mhawkinsbasis**
…o separate interface classes. Created a new AbstractControl class which can be used for classes that aren't necessarily built on DWC controls. Reconfigured which controls implement which interfaces, and implemented default behavior in the Abstract classes for the interfaces they implement. Added a catchUp method that was missing in Label, and went through many of the controls to change their implementation of Boolean checks to conform to best practices

### CODE REVIEW

File: src/main/java/org/dwcj/controls/AbstractControl.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+import org.dwcj.interfaces.IControl;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.basis.bbj.proxies.sysgui.BBjControl;
+import com.basis.startup.type.BBjException;
+
+public class AbstractControl implements IControl {
```
Comment (hyyan) [inline src/main/java/org/dwcj/controls/AbstractControl.java:11]: The class need to be `abstract`

File: src/main/java/org/dwcj/interfaces/IControl.java [hunk 2]
```diff
@@ -3,83 +3,36 @@
 public interface IControl {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IControl.java:3]: Rename to `Control`

File: src/main/java/org/dwcj/interfaces/IAttributable.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package org.dwcj.interfaces;
+
+public interface IAttributable {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IAttributable.java:3]: These adjectives sound unneutral. We should come up with something more  neutral. for instance:

- `HasAttribute` 
- `SupportsAttribute `

This apply of other interfaces too.

File: src/main/java/org/dwcj/interfaces/IControlTextable.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package org.dwcj.interfaces;
+
+public interface IControlTextable {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IControlTextable.java:3]: What about the following for a name ?

- 'HasText'
- 'SupportsText'

File: src/main/java/org/dwcj/interfaces/ICssClassable.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package org.dwcj.interfaces;
+
+public interface ICssClassable {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ICssClassable.java:3]: What about the following for a name ?

- 'HasClassName'
- 'SupportsClassName'

File: src/main/java/org/dwcj/interfaces/ICssClassable.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    
+    /**
+     * Adds a CSS class to the list of CSS classes for the control.
+     * 
+     * @return The control itself
+     */
+    public ICssClassable addClass(String selector);
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ICssClassable.java:10]: rename to `addClassName`

File: src/main/java/org/dwcj/interfaces/ICssClassable.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+    /**
+     * Removes a CSS class from the list of CSS classes for the control.
+     *
+     * @return The control itself
+     */
+    public ICssClassable removeClass(String selector);
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ICssClassable.java:17]: rename to `removeClassName`

File: src/main/java/org/dwcj/interfaces/ICssStylable.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package org.dwcj.interfaces;
+
+public interface ICssStylable {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ICssStylable.java:3]: What about the following for a name ?

- 'HasStyle'
- 'SupportsStyle'

File: src/main/java/org/dwcj/interfaces/IEnableable.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package org.dwcj.interfaces;
+
+public interface IEnableable {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IEnableable.java:3]: How about the following for a name?

`HasEnabled`
`SupportsEnabled`

File: src/main/java/org/dwcj/interfaces/ITextHighlightable.java [hunk 10]
```diff
@@ -1,6 +1,6 @@
 package org.dwcj.interfaces;
 
-public interface ITextControl {
+public interface ITextHighlightable {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ITextHighlightable.java:3]: How about the following for a name?

`TextHighlightable`

File: src/main/java/org/dwcj/interfaces/ITooltipable.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package org.dwcj.interfaces;
+
+public interface ITooltipable {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ITooltipable.java:3]: How about the following for a name?

`HasTooltip`
`SupportsTooltip`

File: src/main/java/org/dwcj/interfaces/IVisible.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
+package org.dwcj.interfaces;
+
+public interface IVisible {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IVisible.java:3]: How about the following for a name?

`HasVisible`
`SupportsVisible`

File: src/main/java/org/dwcj/interfaces/IAttributable.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IAttributable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/IComputedStylable.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IComputedStylable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/IControl.java (hunk 2 above)
Comment (hyyan) [reply src/main/java/org/dwcj/interfaces/IControl.java:3]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/IControlTextable.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IControlTextable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/ICssClassable.java [hunk 16]
```diff
@@ -0,0 +1,18 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ICssClassable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/ICssStylable.java [hunk 17]
```diff
@@ -0,0 +1,13 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ICssStylable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/IEnableable.java [hunk 18]
```diff
@@ -0,0 +1,18 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IEnableable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/ITextHighlightable.java [hunk 19]
```diff
@@ -1,6 +1,6 @@
 package org.dwcj.interfaces;
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ITextHighlightable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/ITooltipable.java [hunk 20]
```diff
@@ -0,0 +1,20 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/ITooltipable.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/interfaces/IVisible.java [hunk 21]
```diff
@@ -0,0 +1,18 @@
+package org.dwcj.interfaces;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/IVisible.java:2]: Document the purpose of the interface

File: src/main/java/org/dwcj/controls/AbstractControl.java [hunk 22]
```diff
@@ -0,0 +1,94 @@
+package org.dwcj.controls;
+
+import org.dwcj.interfaces.IControl;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.basis.bbj.proxies.sysgui.BBjControl;
+import com.basis.startup.type.BBjException;
+
```
Comment (hyyan) [inline src/main/java/org/dwcj/controls/AbstractControl.java:10]: Document the purpose of the class

File: src/main/java/org/dwcj/controls/AbstractDwcControl.java [hunk 23]
```diff
@@ -5,17 +5,26 @@
...
 import org.dwcj.interfaces.ITextAlignable;
-import org.dwcj.interfaces.ITextControl;
+import org.dwcj.interfaces.ITextHighlightable;
+import org.dwcj.interfaces.ITooltipable;
+import org.dwcj.interfaces.IVisible;
 
 import java.util.*;
 
 /**
  * The base class for all controls and panels
```
Comment (hyyan) [inline src/main/java/org/dwcj/controls/AbstractDwcControl.java:25]: Document the purpose of the class

**Review by hyyan — CHANGES_REQUESTED** [review]

**Comment by hyyan** [comment]
@mhawkinsbasis We need to come up with better names for the interfaces. It is ok to drop the `I` from the begging and go with something more natural. I left some suggestions. Let me know which variation you think is better.

File: src/main/java/org/dwcj/interfaces/Visible.java [hunk 24]
```diff
@@ -1,6 +1,11 @@
 package org.dwcj.interfaces;
 
-public interface IVisible {
+
+/**
+ * Interface facilitates implementation of behaviors to modify a 
+ * controls visibility on a page
+ */
+public interface Visible {
```
Comment (hyyan) [inline src/main/java/org/dwcj/interfaces/Visible.java:8]: I'm not sure about this. but do you think `HasVisible` is better ?

**Review by hyyan — CHANGES_REQUESTED** [review]

**Comment by hyyan** [comment]
@mhawkinsbasis  Looks good. I asked for few minor changes. 
Regrading the methods documentation, copy them from basis website at the moment

File: src/main/java/org/dwcj/interfaces/Visible.java (hunk 24 above)
Comment (mhawkinsbasis) [reply src/main/java/org/dwcj/interfaces/Visible.java:8]: I thought about HasVisible but that sounded a little off to me for some reason - what about HasVisibility?

**Comment by mhawkinsbasis** [comment]
> @mhawkinsbasis Looks good. I asked for few minor changes. Regrading the methods documentation, copy them from basis website at the moment

Sounds good, I'll change that right now!

File: src/main/java/org/dwcj/interfaces/Visible.java (hunk 24 above)
Comment (hyyan) [reply src/main/java/org/dwcj/interfaces/Visible.java:8]: `HasVisibility` sounds good. Let's go with that, it goes with the pattern we used for other interfaces.

File: src/main/java/org/dwcj/interfaces/HasPopupMenu.java [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * On applicable controls, helps facilitate underlying BBj popup menu
+ * behavior to facilitate implementation of methods
+ * to interact with this behavior.
+ */
+
+public interface HasPopupMenu {
+
+    HasPopupMenu addPopupMenu(int id, String title);
```
[the remark this answers is not in the document]
Comment (hyyan) [reply src/main/java/org/dwcj/interfaces/HasPopupMenu.java:13]: @mhawkinsbasis still missing documentation

File: src/main/java/org/dwcj/interfaces/HasPopupMenu.java [hunk 26]
```diff
@@ -0,0 +1,14 @@
...
+
+    HasPopupMenu removePopupMenu(int id, String title);
```
[the remark this answers is not in the document]
Comment (hyyan) [reply src/main/java/org/dwcj/interfaces/HasPopupMenu.java:15]: @mhawkinsbasis still missing documentation

**Review by hyyan — CHANGES_REQUESTED** [review]

**Review by hyyan — APPROVED** [review]

**Comment by hyyan** [comment]
Looks good, thanks @mhawkinsbasis

**Pull request merged**